First completed version

etsets opened the pull request:
Branch for design review and fixes.

etsets committed:
First completed version

Branch for design review and fixes.

sevangelatos commented on barber.h:
```diff
@@ -1,27 +1,31 @@
 #include <thread>
-#include <chrono>
 
 using namespace std;
```
it is not advisable to use "using" in headers

etsets replied:
:ok:

sevangelatos commented on barber.cpp:
```diff
@@ -1,14 +1,23 @@
+    ,mCustomersNotifier(customersNotifier)
 {
     mBarberThread = std::thread(operating, this);
     mBarberThread.join();
```
So the constructor will never complete as long as this class is "alive"

etsets replied:
:ok:  Implemented and called in a separate function.

sevangelatos commented on barber.cpp:
```diff
@@ -1,14 +1,23 @@
+    ,mBarberNotifier(barberNotifier)
+    ,mCustomersNotifier(customersNotifier)
 {
     mBarberThread = std::thread(operating, this);
```
I don't think you can assign to a thread like that. In my compiler this is an error:

sevangelatos replied:
clang++ -c -pipe -g -std=gnu++11 -Wall -W -fPIC -DQT_QML_DEBUG -I../barber-shop-problem -I. -I/usr/lib/qt/mkspecs/linux-clang -o barber.o ../barber-shop-problem/barber.cpp
../barber-shop-problem/barber.cpp:12:33: error: reference to non-static member function must be called; did you mean to call it with no arguments?
    mBarberThread = std::thread(operating, this);
                                ^~~~~~~~~
                                         ()
../barber-shop-problem/barber.cpp:12:21: error: no matching constructor for initialization of 'std::thread'
    mBarberThread = std::thread(operating, this);
                    ^           ~~~~~~~~~~~~~~~

sevangelatos replied:
GCC:
../barber-shop-problem/barber.cpp: In constructor ‘Barber::Barber(Shop*, Semaphore*, Semaphore*)’:
../barber-shop-problem/barber.cpp:12:48: error: invalid use of non-static member function ‘void Barber::operating()’
     mBarberThread = std::thread(operating, this);
                                                ^
In file included from ../barber-shop-problem/barber.cpp:1:
../barber-shop-problem/barber.h:17:10: note: declared here
     void operating();
          ^~~~~~~~~
make: *** [Makefile:504: barber.o] Error 1

sevangelatos replied:
Under minGW 32bit in windows, it does compile but when trying to run it, it hangs forever.

Using the MS VS2015 compiler, it produces the following errors:
[vc2015_errors.txt](https://github.com/etsets/barber-shop-problem/files/2088261/vc2015_errors.txt)

sevangelatos commented on barber.cpp:
```diff
@@ -1,14 +1,23 @@
     mBarberThread.join();
 }
 
+Barber::~Barber()
```
If there is nothing for the dtor to do, simply omit it. The default one will work just fine.

etsets replied:
:ok:  Removed from both Customer and Barber classes.

sevangelatos commented on barber.cpp:
```diff
@@ -18,7 +27,11 @@ void Barber::operating()
 {
     while (mAlive)
```
How many threads touch mAlive? I don't think you are using it safely. What if the compiler assigns it to a register?

etsets replied:
Just a note after our conversation: you suggested something like sending a null Customer object in order to terminate thread (and eliminate boolean).

sevangelatos commented on barber.cpp:
```diff
@@ -18,7 +27,11 @@ void Barber::operating()
+        pBelongsToShop->removeWaitingCustomer();
+        mBarberNotifier->Signal();
+        cutHair();
+        std::this_thread::sleep_for(std::chrono::seconds(4));
```
If this sleep is simulating cutting hair, I would move it within the cutHair() function

etsets replied:
:ok:

sevangelatos commented on barber.h:
```diff
@@ -1,27 +1,31 @@
+    Barber(Shop* shop, Semaphore* barberNotifier, Semaphore* customersNotifier);
+    ~Barber();
     void terminate();
+    void operating();
```
Does operating need to be public?

etsets replied:
Nope. :ok:

sevangelatos commented on barber.h:
```diff
@@ -1,27 +1,31 @@
-    Barber(Shop* shop);
+    Barber(Shop* shop, Semaphore* barberNotifier, Semaphore* customersNotifier);
+    ~Barber();
     void terminate();
```
How does the client program know when the barber has actually terminated and can exit? Has the barber terminated after terminate() returns?

sevangelatos commented on barber.h:
```diff
@@ -1,27 +1,31 @@
 {
 public:
-    Barber(Shop* shop);
+    Barber(Shop* shop, Semaphore* barberNotifier, Semaphore* customersNotifier);
```
Do you need to pass so many objects to the barber? Who owns the semaphores? Couldn't he access the semaphores through the shop handle?

etsets replied:
:ok:  Implemented accordingly (get semaphores through Shop*).

sevangelatos commented on semaphore.cpp:
```diff
@@ -1,23 +1,24 @@
     mSemaphoreValue--;
-    if (mSemaphoreValue < 0)
+    if (mSemaphoreValue >= 0)
     {
```
What if somebody decrements the value again after you have checked it in the if? What if the value is 1 and 2 threads call wait() concurrently? Will the value become -1?

etsets replied:
After having read cppreference for std::atomic, I had already declared this as 'std::atomic<int> mSemaphoreValue' in order to avoid race conditions (check the .h file). Do you refer to something else ?

etsets replied:
So a lock is still necessary, decrement value at the end, and replace if statement with 'while (mSemaphoreValue <= 0)... lock'

sevangelatos commented on semaphore.cpp:
```diff
@@ -1,23 +1,24 @@
 {
     mSemaphoreValue++;
-    mMutexObject.unlock();
+    mConditionVariable.notify_all();
```
If 10 threads are blocked waiting on the semaphore and Signal() is called once, do you want to unblock all of them?

etsets replied:
:ok: Replaced by notify_one()

sevangelatos commented on semaphore.cpp:
```diff
@@ -1,23 +1,24 @@
     }
+
+    std::unique_lock<std::mutex> lck(mMutexObject);
+    mConditionVariable.wait(lck);
```
What happens in the case of a spurious wake?

> The thread will be unblocked when notify_all() or notify_one() is executed. It may also be unblocked [spuriously](http://en.cppreference.com/w/cpp/thread/condition_variable/wait)

sevangelatos commented on semaphore.h:
```diff
@@ -3,18 +3,21 @@
 using namespace std;
 #include <mutex>
+#include <atomic>
+#include <condition_variable>
```
I think you are correct, a mutex, an int and a condition variable is what you need to implement a semaphore. Unfortunately your implementation is not correct but you are close.

sevangelatos commented on semaphore.h:
```diff
@@ -3,18 +3,21 @@
 
 private:
-    int mSemaphoreValue;
+    std::atomic<int> mSemaphoreValue;
```
I don't think that using std::atomic makes sense here.

etsets replied:
I have found in similar examples, that such values are declared as atomic. Why do you think its usage does not make sense ?

etsets replied:
I will mark this conversation as resolved, since the suggestion has already been mentioned above.

sevangelatos replied:
It does not make sense because the value is only accessed or modified when the mutex is locked. So it does not make sense to add the costs associated with atomic since the mutex guarantees that the value will be consistent.

sevangelatos commented on shop.cpp:
```diff
@@ -3,24 +3,57 @@
-    ,mBarberSemaphore(0)
-    ,mCustomersSemaphore(0)
 {
+    mBarberSemaphore = new Semaphore(1);
```
Why allocate all semaphores on the heap?

etsets replied:
:ok:

sevangelatos commented on shop.cpp:
```diff
@@ -3,24 +3,57 @@
+
+Shop::~Shop()
+{
+    if (mTheBarber != nullptr)
```
You don't need to check pointers against null when calling [delete](http://en.cppreference.com/w/cpp/memory/new/operator_delete).
> In all cases, if ptr is a null pointer, the standard library deallocation functions do nothing

etsets replied:
:ok: Done

sevangelatos commented on shop.cpp:
```diff
@@ -3,24 +3,57 @@
+{
+    if (mTheBarber != nullptr)
+    {
+        delete mTheBarber;
```
How can you be certain that the barber thread has exited before you delete the barber object?

sevangelatos commented on shop.cpp:
```diff
@@ -3,24 +3,57 @@
+void Shop::newCustomerArrives(std::string customerName)
+{
+    mShopMutex.lock();
+    mTheCustomers.emplace_back(new Customer(customerName, this, mBarberSemaphore, mCustomersSemaphore));
```
How are the customer objects deleted?

sevangelatos replied:
Who owns the customers? Is it reasonable for the shop to create the customers?

etsets replied:
You are right, it is more reasonable that the Customers are externally created (in main.cpp) and then added to the Shop. In this way, main should also handle Customers deletion.

sevangelatos commented on shop.cpp:
```diff
@@ -3,24 +3,57 @@
+
+bool Shop::emptySeatsExist()
+{
+    mShopMutex.lock();
```
You can use a std::lock_guard here.

etsets replied:
:ok: Done in all cases used.

sevangelatos reviewed (changes requested):
Overall, I don't think this version compiles. You can apply the changes to the .pro file that I proposed to the other merge request and make sure that is compiles and is warning-clean.

etsets committed:
Code review improvements part 1

Code review improvements part 1

etsets committed:
Code review suggestions part 2

Code review suggestions part 2

etsets committed:
Code review suggestions part 3

Code review suggestions: proper handling of thread exit check before destroying corresponding objects. Move customers' handling to main function.

etsets commented:
Implementation after review regarding threads termination has been based on the following article: [C++11 : How to Stop or Terminate a Thread](http://thispointer.com/c11-how-to-stop-or-terminate-a-thread/) . Debugging and testing will follow. Please, review so far, so that I am sure that I have clearly understood what you meant.

sevangelatos commented on barber.h:
```diff
@@ -1,27 +1,32 @@
+
     Shop* pBelongsToShop;
+    Semaphore* mBarberNotifier;
+    Semaphore* mCustomersNotifier;
```
Is there any benefit to keep here both the shop and the shop's semaphores? You can always get the semaphores from the shop.

etsets replied:
Done :ok:

sevangelatos commented on barber.cpp:
```diff
@@ -1,28 +1,50 @@
+bool Barber::stopRequested()
+{
+    // checks if value in future object is available
+    if (futureObj.wait_for(std::chrono::milliseconds(0)) == std::future_status::timeout)
```
What is the point of future and "waiting" if you are not going to actually wait? You are still essentially checking a boolean here.

etsets replied:
This solution of a promise is suggested in order to terminate a thread. See the link I posted above, in my comment right after my latest push (http://thispointer.com/c11-how-to-stop-or-terminate-a-thread/). Consider that the shop should close at some point, so the barber should stop operating. Previously you suggested that I should remove the boolean value since it was not safe and it was in general a bad pattern, and probably send a NULL object in order to stop Barber's operation.

sevangelatos replied:
The future construct is thread safe, as an atomic Boolean would also be thread safe. The reason I don't like this approach either is that it is based on polling. Previously, you were polling the value of a Boolean flag. Now you are polling a future. The problem comes from the fact that the barber must react to both customers appearing and to a termination request. With the current design, he can't wait forever on one event because the other might happen. So he has to wait a little bit here (clients semaphore) and a little bit there (termination future). This checking of one channel and then checking the other is a form of polling that keeps him busy. I think, what I mean will become clear when you start running the program and you see it in action.

sevangelatos commented on barber.cpp:
```diff
@@ -1,28 +1,50 @@
+        mCustomersNotifier->Wait();
+        pBelongsToShop->removeWaitingCustomer();
+        mBarberNotifier->Signal();
+        cutHair();
```
I would expect here to have a concept of which client's hair he is cutting.

etsets replied:
Will try to find a way. For the moment each client types its own name.

sevangelatos commented on barber.h:
```diff
@@ -1,27 +1,32 @@
     Barber(Shop* shop);
-    void terminate();
+    void joinThread();
+    bool stopRequested();
```
Could this be const? Could it be private?

etsets replied:
Declared as private :ok:

sevangelatos commented on customer.cpp:
```diff
@@ -1,28 +1,66 @@
 {
-    getHaircut();
-    std::this_thread::sleep_for(std::chrono::seconds(1));
+    while (!stopRequested())
```
The client has a linear lifetime. No need for loop and stop request. I think he goes in, if he finds a chair waits, gets a haircut and vanishes. If he doesn't find a chair, he just vanishes. Why the loop?

etsets replied:
Your question is absolutely logical. My thought was that it is possible that some customers could be waiting (threads active but not served yet) when barber gets a request to stop. In this case one option could be to force the remaining customers to terminate. Another option could be to serve them, do not allow any more to be added and then the barber can be terminated. Do you agree with this second concept ?

sevangelatos replied:
I agree. I think serving all the clients sitting in the shop before shutting down makes total sense.

sevangelatos commented on customer.h:
```diff
@@ -1,23 +1,33 @@
+    Semaphore* mBarberNotifier;
+    Semaphore* mCustomersNotifier;
+
+    bool mTerminated;
```
Is this accessed by many threads? Is it safe? Maybe this would be a good candidate for std::atomic. But I am not sure whether it is actually needed.

etsets replied:
It is used in Shop::clearFirstTerminatedCustomerFound() function (accessed only in Shop, thus the main thread) in order to remove a Customer object that has been served (in order to avoid an increasing vector) from std::vector that holds references to created Customers.

etsets replied:
Declared it as atomic.

sevangelatos commented on semaphore.cpp:
```diff
@@ -1,23 +1,30 @@
-    mSemaphoreValue--;
-    if (mSemaphoreValue < 0)
+    std::unique_lock<std::mutex> lck(mMutexObject);
+    while (mSemaphoreValue < 0)
```
What if it's equal to 0?

etsets replied:
Will change condition to while (mSemaphoreValue < 0) :ok:

sevangelatos commented on semaphore.h:
```diff
@@ -1,20 +1,22 @@
 
 private:
-    int mSemaphoreValue;
+    std::atomic<int> mSemaphoreValue;
```
Do you need this to be atomic? Why?

sevangelatos commented on shop.cpp:
```diff
@@ -1,26 +1,77 @@
+    ,mBarberSemaphore(1)
     ,mCustomersSemaphore(0)
 {
+    mTheBarber = new Barber(this);
```
I think you are overusing pointers and new. If the barber is owned by the shop and lives as long as the shop lives, why not have him as a member instead of keeping a pointer?

etsets replied:
:ok:

sevangelatos commented on shop.cpp:
```diff
@@ -1,26 +1,77 @@
+Shop::~Shop()
+{
+    delete mTheBarber;
+    mTheCustomers.clear();
```
Who is deleting the customers?

etsets replied:
Call is invoked by the Shop. You can see main function.

sevangelatos commented on shop.cpp:
```diff
@@ -1,26 +1,77 @@
+    }
+    *it = mTheCustomers.back();
+    c->joinThread();
+    delete c;
```
This is overly complex and I suspect buggy. You assume that immediately after the "remove waiting customer" that customer will have terminated? What if he is still getting a haircut? Also, why don't you remove him from mTheCustomers or at the very least make the pointer that points to him null? If you keep deleted customers in the list, the second time this function runs, you will ask a deleted customer if he has terminated. You will touch freed memory and boom :-(

etsets replied:
In order to delete an item from vector, I check first if Customer has been terminated (the member value that becomes true after hairccut is over). I followed a suggestion of yours regarding removing items from vectors, but I will simplify it.

sevangelatos replied:
I understand what you intended to do but it is buggy. Currently the size of mCustomers does not decrease.

etsets replied:
I have simplified deletion, pushed a few minutes ago.

sevangelatos commented on shop.cpp:
```diff
@@ -1,26 +1,77 @@
+        c = (*it);
+        break;
+    }
+    *it = mTheCustomers.back();
```
Are you trying here to remove the deleted customer from the list? I think it will not work.

etsets replied:
Yes, this is my purspose. I will simplify it as I mentioned above. :ok:

sevangelatos commented:
It's improving. It's actually looking like we have a skeleton for a first solution. 

Just a note: Prefer not to commit code that does not compile. It is very helpful to get in the habit of keeping the repository in an always compiling state. I find that helpful for 3 reasons:

1) If you do that, you can then use "git bisect" to find which commit introduced a bug automatically, by using binary search.
2) It is more CI friendly.
3) It is also more polite if you are working in a team.

Cheers,
Spiros

etsets committed:
Code review suggestions part 4

Review suggestions part 4

etsets committed:
Debug helpful messages and initialization

Initialization of mWaitingCustomers because its value was not defined and caused instable behavior. Added useful messages for debugging (before being finally able to run properly debugger with breakpoints) .

etsets committed:
Continued debugging and fixed some initialization bugs that caused crashing

etsets committed:
Removed mutex that caused blocking of customer code. Removed customer thread.

Commit:
Fix warning in customer constructor.

etsets committed:
Made barber thread unique_ptr

etsets committed:
Added Barber::start()

etsets committed:
Refactored shop to terminate barber by giving him null customer.

etsets committed:
Destroy Barber on shutdown

etsets committed:
Changed .pro file to compile in linux

etsets committed:
Run each customer to it's own thread.

etsets committed:
Removed variable for waiting customers. Better simulation in order to balk.
